File: src/main/java/dev/isxander/evergreenhud/elements/type/TextElement.java
/*
 * Copyright (C) isXander [2019 - 2021]
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/gpl-3.0.en.html
 *
 * If you have any questions or concerns, please create
 * an issue on the github page that can be found here
 * https://github.com/isXander/EvergreenHUD
 *
 * If you have a private concern, please contact
 * isXander @ business.isxander@gmail.com
 */

package dev.isxander.evergreenhud.elements.type;

import dev.isxander.evergreenhud.settings.impl.BooleanSetting;
import dev.isxander.evergreenhud.settings.impl.EnumSetting;
import dev.isxander.evergreenhud.settings.impl.IntegerSetting;
import dev.isxander.evergreenhud.settings.impl.StringSetting;
import dev.isxander.xanderlib.utils.HitBox2D;
import dev.isxander.xanderlib.utils.Resolution;
import dev.isxander.xanderlib.utils.json.BetterJsonObject;
import lombok.Getter;
import net.minecraft.client.gui.ScaledResolution;

import java.awt.*;

public abstract class TextElement extends BackgroundElement {

    @Getter private StringSetting titleText;
    @Getter private BooleanSetting brackets;
    @Getter private EnumSetting<TextMode> textMode;
    @Getter private BooleanSetting chroma;
    @Getter private EnumSetting<Alignment> alignment;
    private IntegerSetting textR;
    private IntegerSetting textG;
    private IntegerSetting textB;

    /**
     * @return the default display title
     */
    public abstract String getDefaultDisplayTitle();

    /**
     * Gets the box around the text
     *
     * @param gl the gl scale
     * @param sizeScale the modified scale
     * @return hitbox for rendering & gui
     */
    @Override
    public HitBox2D calculateHitBox(float gl, float sizeScale) {
        HitBox2D hitbox = null;
        ScaledResolution res = Resolution.get();

        float width = getHitBoxWidth() * sizeScale;
        float height = getHitBoxHeight() * sizeScale;

        float top = getPaddingTop().get() * sizeScale;
        float bottom = getPaddingBottom().get() * sizeScale;
        float left = getPaddingLeft().get() * sizeScale;
        float right = getPaddingRight().get() * sizeScale;

        float x = getPosition().getRawX(res) / gl;
        float y = getPosition().getRawY(res) / gl;

        switch (getAlignment().get()) {
            case RIGHT:
                hitbox = new HitBox2D(x - (width / sizeScale) - left, y - top, width + left + right, height + top + bottom);
                break;
            case CENTER:
                hitbox = new HitBox2D(x - (width / 2f) - left, y - top, width + left + right, height + top + bottom);
                break;
            case LEFT:
                hitbox = new HitBox2D(x - left, y - top, width + left + right, height + top + bottom);
                break;
        }
        return hitbox;
    }

    @Override
    protected void registerDefaultSettings() {
        super.registerDefaultSettings();

        addSettings(brackets = new BooleanSetting("Brackets", "Display", "If there are square brackets before and after the text.", false));
        addSettings(titleText = new StringSetting("Title", "Display", "What is displayed before or after the value.", getDefaultDisplayTitle()));
        addSettings(textR = new IntegerSetting("Text Red", "Color", "How much red is in the color of the text.", 255, 0, 255, ""));
        addSettings(textG = new IntegerSetting("Text Green", "Color", "How much green is in the color of the text.", 255, 0, 255, ""));
        addSettings(textB = new IntegerSetting("Text Blue", "Color", "How much blue is in the color of the text.", 255, 0, 255, ""));
        addSettings(chroma = new BooleanSetting("Chroma Text", "Color", "If the color of the text is a multicolored mess.", false));
        addSettings(textMode = new EnumSetting<>("Text Mode", "Display", "How should the text be rendered.", TextMode.SHADOW));
        addSettings(alignment = new EnumSetting<>("Alignment", "Display", "When the text grows or shrinks in size, which way the element will move.", Alignment.LEFT));
    }

    @Override
    public void loadJsonOld(BetterJsonObject root) {
        if (!root.optBoolean("title", true)) {
            titleText.set("");
        }
        brackets.set(root.optBoolean("brackets", false));
        chroma.set(root.optBoolean("chroma", false));
        textMode.set((root.optBoolean("shadow", true) ? TextMode.SHADOW : TextMode.NORMAL));

        if (root.has("alignment")) {
            Alignment alignment = Alignment.values()[root.optInt("alignment", 0)];

            if (alignment == Alignment.LEFT) {
                alignment = Alignment.RIGHT;
            } else if (alignment == Alignment.RIGHT) {
                alignment = Alignment.LEFT;
            }

            this.alignment.set(alignment);
        }
        if (root.has("align")) {
            alignment.set(Alignment.values()[root.optInt("align", 0)]);
        }

        BetterJsonObject textColor = new BetterJsonObject(root.get("textColor").getAsJsonObject());
        textR.set(textColor.optInt("r"));
        textG.set(textColor.optInt("g"));
        textB.set(textColor.optInt("b"));

        super.loadJsonOld(root);
    }
    public Color getTextColor() {
        return new Color(textR.get(), textG.get(), textB.get(), 255);
    }

    public void setTextColor(int r, int g, int b) {
        textR.set(r);
        textG.set(g);
        textB.set(b);
    }

    public enum TextMode {
        NORMAL,
        SHADOW,
        BORDER
    }

    public enum Alignment {
        LEFT,
        CENTER,
        RIGHT
    }

}

File: src/main/kotlin/dev/isxander/evergreenhud/elements/type/TextElement.kt
/*
 * EvergreenHUD - A mod to improve your heads-up-display.
 * Copyright (c) isXander [2019 - 2022].
 *
 * This work is licensed under the GPL-3 License.
 * To view a copy of this license, visit https://www.gnu.org/licenses/gpl-3.0.en.html
 */

package dev.isxander.evergreenhud.elements.type

import dev.isxander.settxi.impl.*
import dev.isxander.evergreenhud.event.ClientTickEvent
import dev.isxander.evergreenhud.settings.color
import dev.isxander.evergreenhud.utils.Color
import dev.isxander.evergreenhud.utils.HitBox2D

abstract class TextElement
@JvmOverloads constructor(title: String, cacheTime: Int = 5) : BackgroundElement() {
    var brackets by boolean(false) {
        name = "Brackets"
        category = "Text"
        description = "Text is displayed within [square brackets]"
    }

    var title by string(title) {
        name = "Title"
        category = "Text"
        description = "What is displayed before or after the actual value."
    }

    var textColor by color(Color.white) {
        name = "Color"
        category = "Text"
        description = "The color of the text."
        canHaveChroma = true
    }

    var textStyle by option(TextStyle.SHADOW) {
        name = "Text Style"
        category = "Text"
        description = "What style text is rendered in."
    }

    var alignment by option(Alignment.LEFT) {
        name = "Alignment"
        category = "Text"
        description = "How the text is aligned."
    }

    var cacheTime by int(cacheTime) {
        name = "Text Cache"
        category = "Text"
        description = "How many client ticks to wait before re-calculating the value. (20 ticks = 1 second)"
        range = 0..20
    }

    private val clientTickEvent by event<ClientTickEvent> {
        clientTicks++
        if (clientTicks > cacheTime)
            clientTicks = 0
    }
    var clientTicks = 0

    override fun calculateHitBox(scale: Float): HitBox2D {
        val width = hitboxWidth.coerceAtLeast(minWidth) * scale
        val height = hitboxHeight.coerceAtLeast(minHeight) * scale

        val top = paddingTop * scale
        val bottom = paddingBottom * scale
        val left = paddingLeft * scale
        val right = paddingRight * scale

        val x = position.rawX
        val y = position.rawY

        return when (alignment) {
            Alignment.LEFT -> HitBox2D(x - left, y - top, width + left + right, height + top + bottom)
            Alignment.RIGHT -> HitBox2D(x - (width / scale) - left, y - top, width + left + right, height + top + bottom)
            Alignment.CENTER -> HitBox2D(x - (width / 2f) - left, y - top, width + left + right, height + top + bottom)
            else -> throw IllegalStateException("Failed to parse alignment.")
        }
    }

    object TextStyle : OptionContainer() {
        val NORMAL = option("Normal", "Just plain text.")
        val SHADOW = option("Shadow", "Text with Minecraft's shadow. This can be modified with mods like patcher.")
        val BORDER = option("Border", "Adds a black border around normal text. (Can cause increased amount of lag)")
    }

    object Alignment : OptionContainer() {
        val LEFT = option("Left", "When the length of text is increased, it expands to the right.")
        val RIGHT = option("Right", "When the length of text is increased, it expands to the left.")
        val CENTER = option("Center", "When the length of text is increased, it expands both left and right equally.")
    }
}
